server/mqtt.go

```diff
@@ -3837,37 +3921,44 @@ func (c *client) mqttHandlePubRetain() {
 	// but should still be delivered as a normal message.
 	if pp.sz == 0 {
 		if seqToRemove := asm.handleRetainedMsgDel(key, 0); seqToRemove > 0 {
-			asm.deleteRetainedMsg(seqToRemove)
-			asm.notifyRetainedMsgDeleted(key, seqToRemove)
+			go func() {
```

Comment:
2/5 `mqttHandlePubRetain` is opportunistic to begin with, as I understand it we only needed to block to get the sequence numbers updated in the session cache; can be done in a goroutine, and unblock `c.processIncomingMsg`?

0/5 doing this with an unbounded number of goroutines

Reply:
Also, the order of execution will be different here; perhaps best to funnel through a channel/queue.

Reply:
Removed. 

Going back to processing messages coming from ROUTER and such, isn't it possible that there are mqtt (retain=true) messages coming over from ROUTER, and wouldn't we be then holding the ROUTER client read loop? That'd seem bad?.. Or am I confused and this is only invoked for locally received MQTT messages?